File: 0797-all-paths-from-source-to-target/0797-all-paths-from-source-to-target.py
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        visited = [0 for _ in range(len(graph))]
        answer =[]
        
        def dfs(node, path):
            flag = False
            if path[-1] == len(graph) - 1:
                answer.append(path)
                return 
            
            for _next in graph[node]:
                if visited[_next] == 0:
                    visited[_next] = 1
                    dfs(_next, path + [_next])
                    visited[_next] = 0
                    flag = True
                    
                    
        visited[0] = 1
        dfs(0, [0])
            
        return answer

File: 0797-all-paths-from-source-to-target/0797-all-paths-from-source-to-target.rs
impl Solution {
    pub fn all_paths_source_target(graph: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let n = graph.len();
        let mut answer: Vec<Vec<i32>> = vec![];
        let mut visited = vec![0;n];
        
        fn dfs(graph:&Vec<Vec<i32>>,  node:i32, path:Vec<i32>, n: i32, answer: &mut Vec<Vec<i32>>, visited:&mut Vec<i32> ) {
            if node == n as i32 - 1 {
                answer.push(path);
                return;
            }
            for _next in &graph[node as usize] {
                if visited[*_next as usize] == 0 {
                    visited[*_next as usize] = 1;
                    let mut newPath = path.clone();
                    newPath.push(*_next); 
                    dfs(graph, *_next, newPath, n, answer, visited);
                    visited[*_next as usize] = 0;
                }
            }
        };
        dfs(&graph, 0, vec![0], n as i32, &mut answer, &mut visited);
        answer
    }
}